### PR TITLE
feat(databaseChange): Automatically adjust the timeout if the database change task involves time-consuming index change operations

### LINKED ISSUE #772: [Feature]: Auto-recognize time-consuming SQL and prompts when creating database change task
### Is your feature request related to a problem?

I create a database change task in which create a index for big table and consume long time (> 48h). 
Then the task will timeout and failed, the index may rollback.

### Describe the solution you'd like

May create a new SQL check rule to delect create index operation and prompt user to increase the task timeout?

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PeachThinking**
#### What type of PR is this?
type-feature
module-database change

#### What this PR does / why we need it:
Automatically adjust the timeout if the database change task involves time-consuming index change operations.
1. create `IndexChangeTimeConsumingExists` to check if sqls involve time-consuming operations like creating index and modifying primary key. This SQL check result should not be displayed to the user, so this check result will be filtered out at the end.
2. if involves, `SqlCheckTaskResult#timeConsumingSqlExists` will be true, and `DatabaseChangeResult#isAutoModifyTimeout` will be true.

`ServiceTaskInstanceRepository#findByFlowInstanceId` is a bug, it would return multiple rows of data，But this method is not actually called. Delete this method.

#### Which issue(s) this PR fixes:

Fixes #772 

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DatabaseChangeRuntimeFlowableTaskCopied.java [hunk 1]
```diff
@@ -218,4 +232,26 @@ private JobDefinition buildJobDefinition(DelegateExecution execution) {
                 .build();
     }
 
+    private void modifyTimeoutIfInvolveIndexChange(DelegateExecution execution, DatabaseChangeParameters parameters,
+            DatabaseChangeTaskParameters taskParameters, TaskEntity taskEntity) {
+        Long preCheckTaskId = FlowTaskUtil.getPreCheckTaskId(execution);
+        TaskEntity preCheckTask = taskRepository.findById(preCheckTaskId)
```
Comment (krihy) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DatabaseChangeRuntimeFlowableTaskCopied.java:238]: why not use TaskService#detail to query

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DatabaseChangeRuntimeFlowableTask.java [hunk 2]
```diff
@@ -193,4 +208,23 @@ private DatabaseChangeThread generateOdcAsyncTaskThread(Long taskId, DelegateExe
         return returnVal;
     }
 
+    private void modifyTimeoutIfInvolveIndexChange(DelegateExecution execution, DatabaseChangeParameters parameters) {
+        Long taskId = FlowTaskUtil.getTaskId(execution);
+        Long preCheckTaskId = FlowTaskUtil.getPreCheckTaskId(execution);
+        TaskEntity preCheckTask = taskRepository.findById(preCheckTaskId)
```
Comment (krihy) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DatabaseChangeRuntimeFlowableTask.java:214]: same as above

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/model/SqlCheckTaskResult.java [hunk 3]
```diff
@@ -48,9 +48,10 @@ public class SqlCheckTaskResult implements Serializable, FlowTaskResult {
     private String error;
     private String fileName;
     private List<CheckResult> results = new ArrayList<>();
+    private boolean involveIndexChange;
```
Comment (krihy) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/model/SqlCheckTaskResult.java:51]: suggest: involveIndexChange -> isInvolveIndexChanged

File: server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/rule/IndexChangeTimeConsumingExists.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import lombok.NonNull;
+
+/**
+ * {@link IndexChangeTimeConsumingExists}
+ *
+ * @author jingtian
+ * @date 2024/2/19
+ * @since ODC_release_4.2.4
+ */
+public class IndexChangeTimeConsumingExists implements SqlCheckRule {
```
Comment (krihy) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/rule/IndexChangeTimeConsumingExists.java:43]: suggest add test case for IndexChangeTimeConsumingExists

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DatabaseChangeRuntimeFlowableTaskCopied.java (hunk 1 above)
Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/DatabaseChangeRuntimeFlowableTaskCopied.java]: you are right! I have modified the code

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/model/SqlCheckTaskResult.java (hunk 3 above)
Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/task/model/SqlCheckTaskResult.java]: -> involveTimeConsumingSql

File: server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/SqlCheckService.java [hunk 5]
```diff
@@ -94,7 +94,7 @@ public List<CheckResult> check(@NotNull ConnectionSession session,
     }
 
     public List<CheckViolation> check(@NotNull Long environmentId, @NonNull String databaseName,
-            @NotNull List<OffsetString> sqls, @NotNull ConnectionConfig config) {
+            @NotNull List<OffsetString> sqls, @NotNull ConnectionConfig config, List<SqlCheckRule> defaultRules) {
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/SqlCheckService.java:97]: How to understand the `defaultRules`? Additional rules? Why not create a new method such as `public List<CheckViolation> check(@NotNull Long environmentId, @NonNull String databaseName, @NotNull List<OffsetString> sqls, @NotNull ConnectionConfig config, List<SqlCheckRule> rules)`. In this method, you can specify what rules you want to use.

Comment (PeachThinking) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/sqlcheck/SqlCheckService.java]: you are right! Done.

**Review by smallsheeeep — APPROVED** [review]
LGTM

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**